### PR TITLE
[SPARK-16803] [SQL] SaveAsTable does not work when target table is a Hive serde table

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

In Spark 2.0, `SaveAsTable` does not work when the target table is a Hive serde table, but Spark 1.6 works. 

**Spark 1.6**

``` Scala
scala> sql("create table sample.sample stored as SEQUENCEFILE as select 1 as key, 'abc' as value")
res2: org.apache.spark.sql.DataFrame = []

scala> val df = sql("select key, value as value from sample.sample")
df: org.apache.spark.sql.DataFrame = [key: int, value: string]

scala> df.write.mode("append").saveAsTable("sample.sample")

scala> sql("select * from sample.sample").show()
+---+-----+
|key|value|
+---+-----+
|  1|  abc|
|  1|  abc|
+---+-----+
```

**Spark 2.0**

``` Scala
scala> df.write.mode("append").saveAsTable("sample.sample")
org.apache.spark.sql.AnalysisException: Saving data in MetastoreRelation sample, sample
 is not supported.;
```

So far, we do not plan to support it in Spark 2.1 due to the risk. Spark 1.6 works because it internally uses insertInto. But, if we change it back it will break the semantic of saveAsTable (this method uses by-name resolution instead of using by-position resolution used by insertInto). More extra changes are needed to support `hive` as a `format` in DataFrameWriter.

Instead, users should use insertInto API. This PR corrects the error messages. Users can understand how to bypass it before we support it in a separate PR.
### How was this patch tested?

Test cases are added

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @cloud-fan

**Comment by cloud-fan** [comment]
LGTM

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala [hunk 1]
```diff
@@ -413,6 +413,40 @@ class MetastoreDataSourcesSuite extends QueryTest with SQLTestUtils with TestHiv
...
+  test("saveAsTable(CTAS) using overwrite when the target table is Hive serde") {
+    val tableName = "tab1"
+    withTable(tableName) {
+      sql(s"CREATE TABLE $tableName STORED AS SEQUENCEFILE AS SELECT 1 AS key, 'abc' AS value")
+      val df = sql(s"SELECT key, value FROM $tableName")
+      df.write.mode(SaveMode.Overwrite).saveAsTable(tableName)
+      val tableMeta = spark.sessionState.catalog.getTableMetadata(TableIdentifier(tableName))
+      // When the mode is OVERWRITE, we drop the Hive serde tables and create a data source table
+      // TODO: Based on the definition of OVERWRITE, no change should be made on the table
+      // definition. When recreate the table, we need to create a Hive serde table.
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala]: Sure. Will do it.

**Comment by gatorsmile** [comment]
Thanks! Merging to master/2.1

**Pull request closed without merging**